src/cascadia/QueryExtension/ExtensionPalette.cpp

```diff
@@ -197,48 +201,50 @@ namespace winrt::Microsoft::Terminal::Query::Extension::implementation
+        unique_iter iter{ cmark_iter_new(doc.get()) };
+        cmark_event_type ev_type;
+
+        std::string currentRun{};
+        while ((ev_type = cmark_iter_next(iter.get())) != CMARK_EVENT_DONE)
         {
-            if (!line.empty())
+            const auto node = cmark_iter_get_node(iter.get());
+            const auto nodeType = cmark_node_get_type(node);
+            if (nodeType == CMARK_NODE_TEXT || nodeType == CMARK_NODE_CODE)
             {
-                if (!inCodeBlock && line.find(L"```") == 0)
-                {
-                    inCodeBlock = true;
-                    continue;
-                }
-                if (inCodeBlock && line.find(L"```") == 0)
-                {
-                    inCodeBlock = false;
-                    const auto chatMsg = winrt::make<ChatMessage>(winrt::hstring{ std::move(codeBlock) }, false, true);
-                    messageParts.push_back(chatMsg);
-                    codeBlock.clear();
-                    continue;
-                }
-                if (inCodeBlock)
-                {
-                    if (!codeBlock.empty())
-                    {
-                        codeBlock += L'\n';
-                    }
-                    codeBlock += line;
-                }
-                else
+                // we don't want to create a separate chat message for each text/code node (note that a code node is just an
+                // inline code part, e.g. "...the `-Filter` parameter..."), so just append the raw string here and we'll make
+                // the chat message when we hit a code block or when we end
+                currentRun += cmark_node_get_literal(node);
+            }
+            else if (nodeType == CMARK_NODE_CODE_BLOCK)
+            {
+                // before parsing the code block, append any plaintext we have
+                if (!currentRun.empty())
                 {
-                    const auto chatMsg = winrt::make<ChatMessage>(winrt::hstring{ line }, false, false);
+                    const auto chatMsg = winrt::make<ChatMessage>(winrt::to_hstring(currentRun), false, false);
                     messageParts.push_back(chatMsg);
+                    currentRun.clear();
                 }
+
+                const auto nodeStr = winrt::to_hstring(cmark_node_get_literal(node));
+                // trim the trailing newline
+                std::wstring_view codeView{ nodeStr.c_str(), nodeStr.size() - 1 };
+                const auto chatMsg = winrt::make<ChatMessage>(winrt::hstring{ codeView }, false, true);
+                messageParts.push_back(chatMsg);
             }
         }
+        // append any final plaintext
+        if (!currentRun.empty())
+        {
+            const auto chatMsg = winrt::make<ChatMessage>(winrt::to_hstring(currentRun), false, false);
+            messageParts.push_back(chatMsg);
+            currentRun.clear();
+        }
```

Comment:
Don't like that this is the exact same as lines 227-231 but making a helper function for 3 lines of code seems silly...